Fix occasional db tests failure

rvl opened the pull request:
### Issue Number

Resolves #2064.

### Overview

The DB state machine tests found an extreme case which caused a `too many SQL variables` exception. Unfortunately it was not able to shrink the counterexample without crashing.

- [x] Changes cabal file so that tests are run with a memory limit of 2G.
- [x] Fixes the test case which sometimes fails.

### Comments

- Currently testing with:
   ```
   nix-build -A tests.cardano-wallet-core.unit
   cd lib/core
   while ../../result/bin/unit --match '/Cardano.Wallet.DB.Sqlite/Sqlite State machine tests'; do :; done
   ```
- There is an infinite loop in `Sqlite State machine tests`, it gets stuck somewhere, then takes quite a while to exhaust the heap:
  ```
        There's no checkpoint older than k (+/- 100)
          +++ OK, passed 100 tests.
    Sqlite State machine tests
  1594/100unit: Heap exhausted;
  unit: Current maximum heap size is 2147483648 bytes (2048 MB).
  unit: Use `+RTS -M<size>' to increase it.
   173,889,316,944 bytes allocated in the heap
     9,858,194,528 bytes copied during GC
     2,112,669,264 bytes maximum residency (720 sample(s))
        40,633,672 bytes maximum slop
              2014 MB total memory in use (0 MB lost due to fragmentation)

                                       Tot time (elapsed)  Avg pause  Max pause
    Gen  0     166050 colls,     0 par    7.239s   7.241s     0.0000s    0.0007s
    Gen  1       720 colls,     0 par   2585.867s  2752.903s     3.8235s    6.0089s

    TASKS: 5 (1 bound, 4 peak workers (4 total), using -N1)

    SPARKS: 0(0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

    INIT    time    0.000s  (  0.000s elapsed)
    MUT     time  100.141s  ( 99.704s elapsed)
    GC      time  2593.106s  (2760.143s elapsed)
    EXIT    time    0.002s  (  0.003s elapsed)
    Total   time  2693.249s  (2859.850s elapsed)

    Alloc rate    1,736,438,027 bytes per MUT second

    Productivity   3.7% of total user, 3.5% of total elapsed
  ```
- After disabling shrinking on the state machine tests, the infinite loop turns into a failing test (with quite a large counterexample).
- There is a `too many variables` exception when querying the `tx_meta` table.

rvl committed:
Limit memory usage of cardano-wallet-core:test:unit at 2G

rvl committed:
Try fixing with combineChunked instead

rvl committed:
Fix selectTxHistory with large number of withdrawals

The QuickCheck state machine tests found an extreme case when listing
a transaction history with a large number of withdrawals.

Unfortunately shrinking the counterexample consumed too much memory
and the tests got zapped by the OOM killer.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1378,21 +1379,21 @@ selectTxHistory ti wid minWithdrawal order conditions = do
+                    pure ((`Set.member` txids) . txMetaTxId)
+
+            metas <- filter minWithdrawalPred . fmap entityVal
+                <$> selectList ((TxMetaWalletId ==. wid):conditions) sortOpt
```
Erf. This means that requests like `/transactions?min_withdrawal=1` will have to load the entire transaction history in memory :/  
The case reached by the state-machine is probably a case where there are more than 1000 withdrawals in the wallet right :thinking: ? So we could re-use a similar approach to what was done for fetching inputs / outputs:

```hs
        resolvedInputs <- toOutputMap . fmap entityVal <$>
            combineChunked inputs (\inputsChunk -> selectList
                [TxOutputTxId <-. (txInputSourceTxId <$> inputsChunk)]
                [Asc TxOutputTxId, Asc TxOutputIndex])
```

where `inputs` here would be `txids`. So we can fetch all transactions from the list of `txids` that match the extra conditions, and do it chunk by chunk.

rvl replied:
Yes - not ideal. If using `combineChunked` we also need to sort again after concatenating chunks. But that would be less bad.

KtorZ replied:
I expect most wallets to have a fairly limited amount of withdrawals. And, in case one has many, fetching all transactions first would be way worse.

rvl committed:
Revert "Fix selectTxHistory with large number of withdrawals"

This reverts commit 39c6ee9e931383e80f4dc194bdc51246bf78f679.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1378,21 +1380,24 @@ selectTxHistory ti wid minWithdrawal order conditions = do
-                ) sortOpt
+                    let sortTxId = case order of
+                            W.Ascending -> Desc TxWithdrawalTxId
+                            W.Descending -> Asc TxWithdrawalTxId
```
I am not sure to get the rationale here, surely the extra ordering of the txid has nothing to do with the slot order? So we could pick an arbitrary direction regardless of `order`'s value, right ?

KtorZ committed:
repair GenTxHistory shrinker and generator

  I've made the shrinker a bit less aggressive, such that it only tries to shrink inner values if the list is small enough, otherwise, shrinking large list takes an enormous amount of time. Also, I've tweaked the generator to generate histories of at most 100 transactions, which is enough to still cause the test to fail sometimes, but leads to much faster shrinking and tests altogether. Fix for the test comes in the next commit.

KtorZ committed:
Fix readTxHistory in the context of withdrawals

  There were two issues:
  - We need a double sorting to make it stable, first on the tx id, and
    then on the slot.

  - We need to nub the list of txids fetched from the database
    otherwise, we end up with duplicates in the resulting transaction
    history. I've looked for a 'DISTINCT' clause in persistent for
    SQLite without success ... so nubbing happens in Haskell on the
    fetched ids, which should still be quite okay since even on large
    wallets, the number of withdrawals should be quite small (at most
    one per epoch).

KtorZ reviewed (approved):
I've carried with the PR, fixed the shrinker and the withdrawal fetching which was still causing an issue (see commit)

KtorZ committed:
run stylish haskell, fixing imports declarations
